### PR TITLE
Implement GPU nllLoss and array transfer ops.

### PR DESCRIPTION
**Pull request by dan-zheng**
- Implement `nllLoss` and `nllLoss_grad` kernels.
- Add `GPUArrayFromSeq` expression.
  - This is necessary for initializing the "target" argument of type
    `Array[Int]` on GPU.
- Add array transfer expressions.
  - Array transfer ops are necessary for transferring non-`Tensor` arrays,
    e.g. target labels of type `Rep[Array[Int]]`.
  - Redefine tensor transfer ops in terms of array copy ops.

GPU `nllLoss` and `nllLoss_grad` behave as expected (verified in `lantern-cudnn-nll-loss`).
MNIST CNN works using GPU `nllLoss`.

### CODE REVIEW

**Comment by dan-zheng** [comment]
Ready for review. There are two main changes:
- Implement `nllLoss` and `nllLoss_grad` kernels. These are simple.
- Implement array transfer ops (generalizing tensor data transfer ops).
  - The kernels expect all array arguments (including `target`) to be on the GPU.
  - Array transfer ops allow us to write `target.moveToGPU()`.

MNIST CNN works using GPU `nllLoss`.

File: src/main/scala/lantern/TensorDifferentiation.scala [hunk 1]
```diff
@@ -353,7 +353,7 @@ trait TensorDsl extends DslOps with Diff {
   class BackendCPU protected() extends Backend {
     override def setup() {}
     override def cleanup() {}
-    override def mallocArray[T: Manifest](length: Int): Rep[Array[T]] = NewArray[T](length)
+    override def mallocArray[T: Manifest](length: Rep[Int]): Rep[Array[T]] = NewArray[T](length)
```
[the remark this answers is not in the document]
Comment (dan-zheng) [reply src/main/scala/lantern/TensorDifferentiation.scala]: The MNIST model calls `train.target.moveToGPU(train.tlen)`, where `tlen` is `Rep[Int]`.
To support that, I had to change all `Int` length arguments to `Rep[Int]`.

[the remark this answers is not in the document]
Comment (feiwang3311) [reply src/main/scala/lantern/TensorDifferentiation.scala]: Maybe I misunderstood, but isn't tlen the same as batch_size, which is Int type?
Maybe you tried to move the targets of all training data, which is defined by the file length, thus being Rep[Int]? If that is the case, I think we can just transfer for each batch_size, and still keep the length of arrays Int type.

[the remark this answers is not in the document]
Comment (dan-zheng) [reply src/main/scala/lantern/TensorDifferentiation.scala]: You're right, transferring for each batch size would enable `Int` length argument.
(I optimized prematurely by transferring all of `target` once. It'd be nice to flesh out the data loader API a bit and add a helper that handles transferring.)
